apps/emqx/src/emqx_router.erl

```diff
@@ -213,12 +222,70 @@ do_delete_route(Topic) when is_binary(Topic) ->
 
 -spec do_delete_route(emqx_types:topic(), dest()) -> ok | {error, term()}.
 do_delete_route(Topic, Dest) ->
-    mria_delete_route(get_schema_vsn(), Topic, Dest).
+    mria_delete_route(get_schema_vsn(), Topic, Dest, single).
+
+mria_delete_route(v2, Topic, Dest, Ctx) ->
+    mria_delete_route_v2(Topic, Dest, Ctx);
+mria_delete_route(v1, Topic, Dest, Ctx) ->
+    mria_delete_route_v1(Topic, Dest, Ctx).
+
+-spec do_batch(Batch) -> Errors when
+    %% Operation :: {add, ...} | {delete, ...}.
+    Batch :: #{Route => _Operation :: tuple()},
+    Errors :: #{Route => _Error},
+    Route :: {emqx_types:topic(), dest()}.
+do_batch(Batch) ->
+    Nodes = batch_get_dest_nodes(Batch),
```

Comment:
Side question: I wonder if we can avoid calling `emqx_router_helper:monitor/1` on every `add_route`. 
It seems to be redundant, as emqx_router_helper already receives `{membership,{node,down, Node}}` because it calls `ekka:monitor(membership)`. Alternatively, router_helper can also add local node to `emqx_routing_node` tab once in `init/1` cb.
Performance-wise, it would probably make little sense but the code can be slightly simplified.

Reply:
Yes, it's expected to receive it from `mria_membership` process: https://github.com/emqx/emqx/blob/master/apps/emqx/src/emqx_router_helper.erl#L104

Mria than notifies all monitor callers if node is down:
1. `mria_node_monitor` tracks all connected nodes: https://github.com/emqx/mria/blob/main/src/mria_node_monitor.erl#L78
2. `mria_node_monitor` notifies `mria_membership` when some connected node (either core or replicant) is down: 
https://github.com/emqx/mria/blob/main/src/mria_node_monitor.erl#L133
2. `mria membership` notifies subscribed processes (those that called `ekka/mria:monitor`): https://github.com/emqx/mria/blob/main/src/mria_membership.erl#L509

It looks like emqx_router_helper:monitor/1 is used as a safety measure if underlying ekka/mria node monitor doesn't work (or probably it behaved differently in the past). 
I think we can either stop using `?ROUTING_NODE` and rely solely on ekka/mria in emqx_router_helper, or just write local node to `?ROUTING_NODE` tab once in `emqx_router_helper:init/1`. 
Probably it's worth a separate task/PR, though. :thinking:

Reply:
@zmstone, @ieQu1 may I ask you to have a look at this thread?

Reply:
Thanks for clarifying! My impression was that this thing with `?ROUTING_NODE` table was introduced as a kind of optimization to avoid full table scans for nodes that do not serve any clients whatsoever. And now it looks like this no longer works as intended for quite some time.

Reply:
Yes, it probably was the intention indeed. However, it actually tries to cleanup routes of the down node twice (though during the second call there is nothing more to clean and it much faster):
An example of node down (testing it now in scope of down node routes cleanup optimization):
```
call emqx_router_helper:handle_info({nodedown,'emqx1@127.0.0.1'},#{nodes => ['emqx1@127.0.0.1']}) 
Routes cleaned up in: 6977ms
(<0.2635.0>) call emqx_router_helper:handle_info({membership,{node,down,'emqx1@127.0.0.1'}},#{nodes => []}) 
(<0.2635.0>) call emqx_router_helper:handle_info({nodedown,'emqx1@127.0.0.1'},#{nodes => []}) ({gen_server,
Routes cleaned up in: 411ms
```

Reply:
agree, no need to monitor twice.

Reply:
thanks @zmstone , I'll address it in a separate PR, agreed with @keynslug .